numpy/_core/src/multiarray/ctors.c

```diff
@@ -1615,6 +1622,10 @@ PyArray_FromAny_int(PyObject *op, PyArray_Descr *in_descr,
         assert(cache->converted_obj == op);
         arr = (PyArrayObject *)(cache->arr_or_sequence);
         /* we may need to cast or assert flags (e.g. copy) */
+        if (was_copied_by__array__ == 1 && flags & NPY_ARRAY_ENSURECOPY) {
+            flags = flags & ~NPY_ARRAY_ENSURECOPY;
+            flags = flags | NPY_ARRAY_ENSURENOCOPY;
```

Comment:
This looks wrong, so please add a test:
```
np.asarray(array_like, copy=True, order="F")
```
should just make two copies.  You probably want to make sure that users honors the `dtype` that was passed.  That might be nice, but if it is not very clean to test for, I am not sure we should worry much.

Reply:
Ah right, a copy still needs to be made to convert to `order="F"`. Then I shouldn't add `ENSURENOCOPY` flag but only remove `ENSURECOPY` one.

About honoring `dtype`: I can check if `arr` in cache has the same `dtype` as requested `dtype` variable. I wonder if it's sufficient for comparing dtypes in C-API:
```
PyArray_TYPE(arr) == dtype->type_num
```

Reply:
Then let's maybe skip checking `dtype` after `__array__` call? 
What we currently have is that:
```python
np.arrray(array_like, dtype=new_dtype, copy=False)
```
errors when a `dtype` wasn't honored in `__array__` implementation and
```python
np.arrray(array_like, dtype=new_dtype)
```
get's another copy, same as `order="F"` case.

Reply:
> Then let's maybe skip checking dtype after __array__ call?
For here, that is perfectly fine to me, it is just a mild performance nudge for downstream after all.

However, maybe we should make a new issue.  I am slightly worried that downstream until now always just ignored `dtype=`, and now we will get new bugs with subarray dtypes.

OTOH, I will also say that, I don't expect anyone to explicitly run into it, so in a sense, it is an unrelated bug fix.

Reply:
I updated flags logic here and added a test with:
```python
np.asarray(array_like, copy=True, order="F")
```